### DISCUSSION TITLE
issue: defaultValues not respected after performing yup validation

### DISCUSSION THREAD

**Discussion by Mikilll94** [react-hook-form/react-hook-form#8452] [Q&A]
### Version Number

7.4.1

### Codesandbox/Expo snack

https://codesandbox.io/s/jolly-smoke-rgm4gx?file=/src/App.js:418-431

### Steps to reproduce

1. Open this sandbox
https://codesandbox.io/s/jolly-smoke-rgm4gx?file=/src/App.js:418-431

3. Click the button -> you will see the error: `dateOfBirth is a required field`
4. Click the button again -> you will see the error: `dateOfBirth must be a date type, but the final value was: Invalid Date (cast from the value "").`

As you can see after performing yup validation, the `defaultValues` in `useForm` are not respected and `dateOfBirth` is empty string.

### Expected behaviour

After clicking the button I should always receive error `dateOfBirth is a required field` no matter how many times I clicked the button.

### What browsers are you seeing the problem on?

Chrome

### Relevant log output

N/A

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

**Comment by Moshyfawn** [comment]
It's not bug, per say.

As you've opted in to use the uncontrolled input approach _(the `register` API)_, the source of truth for your current field value is the actual native HTML input's value. Providing the default value as `null` only sets the RHF initial default value and not the HTML input's one. Proving `null` to the native HTML input results in it's value being `"" <empty string>` _(and it's RHF specific behavior, but rather the way native HTML input works)_.
<img width="1035" alt="image" src="https://user-images.githubusercontent.com/16290753/171900333-cb2dba91-1a0b-4b3a-be84-3d6f2cb497f8.png">


As you're trying to get the input value when submitting the form, you'll get an empty `string` instead if your `null` value, which is obviously not a `null` value.

That said, the best  way to go about the `Date` input is to provide the `valueAsDate: true` as a `register` option.
```js
{...register("date", { valueAsDate: true })}
```

**Comment by Mikilll94** [comment]
@Moshyfawn Thanks for you answer. Unfortunately `valueAsDate: true` is not working for me. Please check this CodeSandbox:

https://codesandbox.io/s/serene-kowalevski-2sbfoj?file=/src/App.js

When I click the button `{dateOfBirth: Invalid Date}` is logged to the console.

![image](https://user-images.githubusercontent.com/17653648/171998025-efe3e996-7b4d-44c5-a6bb-d49716a59d78.png)

According to the docs, `null` should be returned:

![image](https://user-images.githubusercontent.com/17653648/171998076-7d03afba-73c4-4806-9d00-a37c519e9c62.png)

https://react-hook-form.com/api/useform/register

**Comment by bluebill1049** [comment]
please validate your own schema, as hook form only provide values for yup to validate and return its result. you can debug in the below codesandbox:

https://codesandbox.io/s/wonderful-dew-p51igr?file=/src/App.js

**Comment by Mikilll94** [comment]
@bluebill1049 
Thanks for the info, but please also check my post above about `valueAsDate` which is not working. I have posted a different CodeSandbox there which does not use yup.

**Comment by bluebill1049** [comment]
It's working just fine. for not valid input value (such as null) which failed with parse data type, it will not return the `date`  type.
<img width="922" alt="Screen Shot 2022-06-04 at 10 34 46 pm" src="https://user-images.githubusercontent.com/10513364/171999115-2a9b1678-1766-4590-9fcb-c201507471dc.png">

**Reply by Mikilll94** [reply]
> for not valid input value (such as null) which failed with parse data type, it will not return the date type.

But the docs are saying **If something goes wrong null will be returned.**

![image](https://user-images.githubusercontent.com/17653648/171999769-8e28c86e-909d-4713-a2b7-1b10bc4abcfc.png)

https://react-hook-form.com/api/useform/register

IMO, this is a bug in the RHF. `null` should be returned instead of string "Invalid date".

**Reply by bluebill1049** [reply]
doc is updated.

**Reply by Mikilll94** [reply]
@bluebill1049 @Moshyfawn  Ok, so how can I implement optional date input with yup? Seems like there is no way to implement it 😞 

I would like to get `null` from input when there is no date instead of "Invalid date".

**Comment by reo777** [comment]
> That said, the best way to go about the Date input is to provide the valueAsDate: true as a register option.

What we should do when we use controlled components?

**Comment by leapful** [comment]
@reo777 

The problem is not from uncontrolled or controlled component. It's from yup schema validation rule. If you want to keep a default empty value for your input type date, just follow the default specification of HTML5 input type date to use empty string then perform yup validation rule conditionally on the input value type.

Working demo: https://codesandbox.io/s/objective-tristan-3y16y6?file=/src/App.js
